htlcswitch/iterator.go

```diff
@@ -85,19 +89,39 @@ type HopIterator interface {
 	// EncodeNextHop encodes the onion packet destined for the next hop
 	// into the passed io.Writer.
 	EncodeNextHop(w io.Writer) error
+
+	// OnionPacket returns the original onion packet used to generate the
+	// secret keys for this instance.
+	OnionPacket() *sphinx.OnionPacket
```

Comment:
Hmm, this is a bit of a violation of the interface. As right now we use a version of this, but a plaintext one within the unit tests. Perhaps instead we can add a method to create an `ErrorEncryptor` directly from it?

Reply:
Creating the error encrypter requires a reference to the `sphinx.Router` that can properly decode the onion packet. I agree it's not perfect, though it seemed preferable than giving every iterator a reference to the router.

Thinking about this some more, maybe expose something like:
```golang
ExtractErrorEncrypter(*sphinx.Router) ErrorEncrypter
```

Internally, this could would just pass in the onion packet to the router, and return the error encrypter. And then the mock implementation can just return an error encrypter directly?

We could also pass in an interface to abstract away the sphinx router, that is solely purposed with handling this decoding.

Does this sound like a good way of approaching it?

Reply:
What about just adding an `ErrorEncrypter() ErrorEncrypter` method to the interface which is just a getter and including an `ErrorEncrypter` field on `sphinxHopIterator`? It should be easy enough because both the router and onion packet are available in `DecodeHopIterator` where it is constructed.